File: 1578-minimumTimeToMakeRopeColorful/minCost.c
/**
 * The idea is to go for a greedy approach. For n (n > 1) consecutive ballos of
 * the same color, keep the one with the maximum neededTime and remove the rest.
 *
 */
int minCost(char *colors, int *neededTime, int neededTimeSize) {
  int totalCost = 0;
  for (int i = 1; i < neededTimeSize; i++) {
    if (colors[i] == colors[i - 1]) {
      if (neededTime[i] < neededTime[i - 1]) {
        totalCost += neededTime[i];
        neededTime[i] = neededTime[i - 1];
      } else {
        totalCost += neededTime[i - 1];
      }
    }
  }
  return totalCost;
}

File: 1578-minimumTimeToMakeRopeColorful/minCost.hs
minCost :: String -> [Int] -> Int
minCost [] [] = 0
minCost [color] [cost] = 0
minCost (color1 : color2 : colors) (cost1 : cost2 : costs)
  | color1 == color2 =
      if cost1 < cost2
        then cost1 + minCost (color2 : colors) (cost2 : costs)
        else cost2 + minCost (color2 : colors) (cost1 : costs)
  | otherwise = minCost (color2 : colors) (cost2 : costs)
